proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,647 @@
+
+(type $CtorFuncWasm (func (result anyref)))
+(func $TextEncoderCtor (import "TextEncoder" "ctor") (type $CtorFuncWasm))
+
+(webidl-func-binding $CtorBinding import $CtorFuncWasm $CtorFuncWebIDL
+  (result (as type=anyref (get 0))))
+(webidl-bind $TextEncoderCtor $CtorBinding)
+```
+
+With these bindings, a WebAssembly module can call `$TextEncoderCtor`, store
+the result in an `anyref` global and then pass that global value as the first
+argument to all subsequent calls to `$encodeInto`.
+
+Lastly, to instantiate this module (without any of the additional ESM features
+that were discussed [above](#JS-API-integration)), the embedding JavaScript would
+populate the import object with functions plucked from the global object stored
+in properties that match the import module/field names above.
+
+```js
+WebAssembly.instantiateStreaming(fetch('encodeInto-example.wasm'), {
+  TextEncoder: {
+    encodeInto: TextEncoder.prototype.encodeInto,
+    ctor: TextEncoder
+  }
+}).then(...);
+```
+
+## FAQ
+
+### Does this proposal add any new low-level capabilities, APIs or backdoors?
+
+No; it just optimizes calls to *existing* APIs that, in the case of the Web,
+are already exposed to JavaScript.
+
+### Will Web IDL Binding Expressions grow to duplicate WebAssembly?
+
+No; the criteria for adding a [binding operator](#binding-operators-and-expressions)
+is that the operator must express something that couldn't be expressed
+(without significant overhead) in WebAssembly.  That's why, e.g., there are no
+numeric conversion binding operators.
+
+### What about bindings for non-Web IDL APIs?
+
+For the case of pre-existing APIs (e.g., Java class libraries or C APIs or RPC
+IDLs), the APIs' interfaces will likely have very different idioms and types
+used in signatures, thereby requiring a binding section with very different
+types and binding operators from the ones presented above.  Attempting to force
+one IDL (with hard backwards-compatibility requirements) to satisfy all these
+disparate use cases would likely lead to a suboptimal outcome for all APIs.
+
+In contrast, having a separate binding section specification for each separate
+idiomatic family of APIs would allow each to be ideally suited and even allow a
+single WebAssembly module, with a single WebAssembly function import to have
+multiple bindings, allowing it to run efficiently in multiple environments (due
+to the ignored-if-unknown default behavior of [custom sections][custom section]).
+
+For the case of new, designed-for-WebAssembly APIs, the API can use core
+WebAssembly value types, avoiding the need for bindings at all, or use Web IDL
+Bindings, gaining a certain amount of Web interoperability "for free".  Based on
+practical experience with Web IDL Bindings, new bindings could be designed in
+the future to optimally handle the use cases that remain, outside the
+constraints of Web APIs' backwards-compatibility.
```

Comment:
So, is WebIDL as described here suitable for non-web use cases? For example, a hypothetical HTTP client:
```
interface Http {
  void sendRequest(HttpRequest req);
};
```

Reply:
Yes, there's nothing that depends here on JavaScript or a browser environment; as long as a host can meaningfully define what Web IDL values mean and how to compile pure-wasm calls to host-calls taking said Web IDL values, you're good.

Reply:
@lukewagner Thank you for clarifying.